### ISSUE TITLE
Avoid unnecessary initialization in nested lazy initializers

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/fenix#15279]
MarcLeclair found a pattern that prevents kotlin's lazy initialization pattern from working as expected in nested lazy initializers when combined with our serviceLocator system:
```kotlin
    val useCases by lazy {
        UseCases(
            search.searchEngineManager,
        )
    }
```

via [code sample](https://searchfox.org/mozilla-mobile/rev/7626be04d759ad7f3e3c050f8b61c524fc1d4fda/fenix/app/src/main/java/org/mozilla/fenix/components/Components.kt#55)

If `useCases` is referenced and thus initialized, `search.searchEngineManager` will unexpectedly be initialized because it's referenced to pass it into the `useCases` constructor, even if it's not actually used in the initializer. In practice, we've identified improvements of up to 60-80ms in just one case where this occurred (in `useCases`).

We should try to identify a generic solution to prevent this anti-pattern for good.

### Initial solution ideas
- we can wrap `Lazy`, or implement our own, to intelligently pass getters rather than the actual object
- maybe we can adapt this one-off solution that ac has used in the past ([definition](https://searchfox.org/mozilla-mobile/rev/7626be04d759ad7f3e3c050f8b61c524fc1d4fda/fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt#250), [usage](https://searchfox.org/mozilla-mobile/rev/7626be04d759ad7f3e3c050f8b61c524fc1d4fda/fenix/app/src/main/java/org/mozilla/fenix/components/Components.kt#40))
- as a last ditch effort, we could implement static analysis to try to prevent this from occurring but it'll add a layer of complexity possibly unlike other solutions

**Comment by mcomella** [comment]
Baron-Severin and I had an interesting discussion: with nested lazy initialization, we can later get a performance issue where, say we're moving through a fragment transaction and trigger a cascading initialization of many components, causing us to drop frames. Given this can be triggered by random code paths, we may not notice this in local development but our users may notice it.

This is no reason not to fix this bug – being in an indeliberate half lazy state is just as bad if not worse – but we should consider a solution for this in the future – I filed https://jira.mozilla.com/browse/FXP-1358 to consider implementing this in the future.

**Comment by mcomella** [comment]
I wrote up a document with my proposed solution: https://docs.google.com/document/d/12O0elEeo5zOBqFwWJ0mZmh4L2d-UNZrnQe5ZrPdj1_o/edit#heading=h.kvxbynlqzbyf I wanted to get feedback before landing it because it affects how the fenix team interacts with components.

**Comment by mcomella** [comment]
I have most of a PR ready to add a lint rule but it's built on changes in this PR https://github.com/mozilla-mobile/fenix/pull/16101 which is blocked by CI from landing. Also, it'd benefit from changes in MarcLeclair's PR https://github.com/mozilla-mobile/fenix/pull/15942

**Comment by mcomella** [comment]
> I have most of a PR ready to add a code to count the number of components we initialize

I want to count the number of components we initialize before we start doing recursive init protection to see how much of an impact the recursive init protection actually makes.

**Comment by mcomella** [comment]
> I have most of a PR ready to add a code to count the number of components we initialize

https://github.com/mozilla-mobile/fenix/pull/16292

After this, we need to add a test for counting the number of components init so we don't regress, then started working on the recursive lazy implementation.

**Comment by mcomella** [comment]
I opened a PR to add `LazyComponent` to android-components: https://github.com/mozilla-mobile/android-components/pull/8874

**Comment by mcomella** [comment]
I put up two PRs as proof-of-concept using `LazyComponent`: https://github.com/mozilla-mobile/android-components/pull/8909 https://github.com/mozilla-mobile/fenix/pull/16410

It was a lot of work to use it so I'm wondering if it's worth the effort to keep going forward. I should probably measure the difference in start up time from this PR.

**Comment by mcomella** [comment]
We're having a discussion on the best approach in the ac PR: https://github.com/mozilla-mobile/android-components/pull/8909#issuecomment-725590234

**Comment by mcomella** [comment]
The `LazyComponent` proof-of-concept turned out to be unnecessary because the `searchEngineManager`, which had the performance problems, was refactored out in a separate PR. See https://github.com/mozilla-mobile/android-components/pull/8909#issuecomment-742762773 for details.

If we find other components with perf issues that we can't optimize out, we can use this approach there. However, there are no actionable items to do right now without further investigation of performance issues during start up so let's close this issue.

**Comment by mcomella** [comment]
To be explicit, we made no perf optimizations in this PR but we came up with an approach to deal with under-performing components.
